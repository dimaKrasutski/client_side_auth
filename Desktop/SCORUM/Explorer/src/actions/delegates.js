import { createActions } from 'redux-actions';

export const {
  preloadActiveWitnesses,
  setActiveWitnesses,
  setCandidateWitnesses,
  preloadWitnessesByVote,
  setWitnesses,
} = createActions(
  'PRELOAD_ACTIVE_WITNESSES',
  'SET_ACTIVE_WITNESSES',
  'SET_CANDIDATE_WITNESSES',
  'PRELOAD_WITNESSES_BY_VOTE',
  'SET_WITNESSES',
);
