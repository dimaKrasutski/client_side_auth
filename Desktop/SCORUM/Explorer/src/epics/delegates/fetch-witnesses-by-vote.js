import { ofType, ActionsObservable } from 'redux-observable';
import { mergeMap } from 'rxjs/operators/mergeMap';
import {
  setWitnesses,
  preloadActiveWitnesses,
  preloadWitnessesByVote,
} from '../../actions/delegates';
import { getWitnessesByVote } from '../../api/scorum';
import { catchGlobalErrorWith } from '../common-operators';
import { UNDEFINED_ERROR_ID } from '../../constants';

export const fetchWitnessesByVoteEpic = action$ => action$.pipe(
  ofType(preloadWitnessesByVote),
  mergeMap(() => getWitnessesByVote()),
  mergeMap(response => (ActionsObservable.of(setWitnesses(response), preloadActiveWitnesses()))),
  catchGlobalErrorWith(UNDEFINED_ERROR_ID),
);
