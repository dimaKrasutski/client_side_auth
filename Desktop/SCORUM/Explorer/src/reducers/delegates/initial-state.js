const initialState = {
  witnesses: [],
  active: [],
  candidates: [],
};

export default initialState;
