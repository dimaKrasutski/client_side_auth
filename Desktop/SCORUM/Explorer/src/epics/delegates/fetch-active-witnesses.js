import { ofType, ActionsObservable } from 'redux-observable';
import { mergeMap } from 'rxjs/operators/mergeMap';
import { map } from 'ramda';
import { setActiveWitnesses, setCandidateWitnesses, preloadActiveWitnesses } from '../../actions/delegates';
import { getActiveWitnesses } from '../../api/scorum';
import { catchGlobalErrorWith } from '../common-operators';
import { UNDEFINED_ERROR_ID } from '../../constants';

export const fetchActiveWitnessesEpic = (action$, store) => action$.pipe(
  ofType(preloadActiveWitnesses),
  mergeMap(() => getActiveWitnesses()),
  mergeMap((response) => {
    const witnesses = map((witness) => {
      const result = witness;
      result.votes = (result.votes / 1000000000).toFixed(9);
      return result;
    }, store.getState().delegates.witnesses);
    const active = witnesses.filter(witness => (response.indexOf(witness.owner) > -1));
    const candidates = witnesses.filter(witness => (response.indexOf(witness.owner) === -1));
    return ActionsObservable.of(setActiveWitnesses(active), setCandidateWitnesses(candidates));
  }),
  catchGlobalErrorWith(UNDEFINED_ERROR_ID),
);
