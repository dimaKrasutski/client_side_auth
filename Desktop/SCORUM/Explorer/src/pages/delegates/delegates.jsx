import React, { Component } from 'react';
import { connect } from 'react-redux';
import { FormattedMessage } from 'react-intl';
import PropTypes from 'prop-types';
import Container from 'reactstrap/lib/Container';
import Row from 'reactstrap/lib/Row';
import Col from 'reactstrap/lib/Col';
import { compose } from 'ramda';
import { preloadWitnessesByVote } from '../../actions/delegates';
import { Main } from '../main';
import { PageTitle } from '../../components/page-title/page-title';
import { WitnessesTable } from './witnesses-table/witnesses-table';

export class DelegatesUI extends Component {
  static propTypes = {
    activeWitnesses: PropTypes.arrayOf(
      PropTypes.shape({
        owner: PropTypes.string.isRequired,
        votes: PropTypes.string.isRequired,
      }),
    ).isRequired,
    candidates: PropTypes.arrayOf(
      PropTypes.shape({
        owner: PropTypes.string.isRequired,
        votes: PropTypes.string.isRequired,
      }),
    ).isRequired,
    preloadWitnessesByVoteAction: PropTypes.func.isRequired,
  };

  componentWillMount() {
    const { preloadWitnessesByVoteAction } = this.props;
    preloadWitnessesByVoteAction();
  }

  render() {
    const { activeWitnesses, candidates } = this.props;
    return (
      <Main>
        <Container className="pt-6 blocks-container">
          <Row>
            <Col xs="12">
              <PageTitle title="explorer.delegates.title" />
              <h4>
                <FormattedMessage id="explorer.delegates.table-active.title" />
              </h4>
              <WitnessesTable witnesses={activeWitnesses} />
              <h4>
                <FormattedMessage id="explorer.delegates.table-candidates.title" />
              </h4>
              <WitnessesTable witnesses={candidates} isCandidates />
            </Col>
          </Row>
        </Container>
      </Main>
    );
  }
}

export const mapStateToProps = ({ delegates }) => ({
  activeWitnesses: delegates.active,
  candidates: delegates.candidates,
});

export const mapDispatchToProps = dispatch => ({
  preloadWitnessesByVoteAction: compose(dispatch, preloadWitnessesByVote),
});

export const Delegates = connect(mapStateToProps, mapDispatchToProps)(DelegatesUI);
