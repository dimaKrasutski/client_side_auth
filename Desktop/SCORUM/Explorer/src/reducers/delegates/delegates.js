import { handleActions } from 'redux-actions';
import initialState from './initial-state';
import { setActiveWitnesses, setCandidateWitnesses, setWitnesses } from '../../actions/delegates';

export default handleActions(
  {
    [setActiveWitnesses]: (state, { payload: active }) => ({
      ...state,
      active,
    }),
    [setCandidateWitnesses]: (state, { payload: candidates }) => ({
      ...state,
      candidates,
    }),
    [setWitnesses]: (state, { payload: witnesses }) => ({
      ...state,
      witnesses,
    }),
  },
  initialState,
);
