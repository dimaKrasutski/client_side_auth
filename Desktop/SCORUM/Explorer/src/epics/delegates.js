import { combineEpics } from 'redux-observable';
import { fetchActiveWitnessesEpic } from './delegates/fetch-active-witnesses';
import { fetchWitnessesByVoteEpic } from './delegates/fetch-witnesses-by-vote';

export const delegatesEpics = combineEpics(fetchActiveWitnessesEpic, fetchWitnessesByVoteEpic);
