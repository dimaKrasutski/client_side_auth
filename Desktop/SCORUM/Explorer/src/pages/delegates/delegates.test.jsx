import React from 'react';
import { DelegatesUI } from './delegates';

describe('<DelegatesUI />', () => {
  test('should match snapshot', () => {
    const wrapper = shallow(
      <DelegatesUI preloadWitnessesByVoteAction={jest.fn()} activeWitnesses={[]} candidates={[]} />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
