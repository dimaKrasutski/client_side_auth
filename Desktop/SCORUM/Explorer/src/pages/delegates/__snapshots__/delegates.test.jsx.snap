// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DelegatesUI /> should match snapshot 1`] = `
<Connect(MainUI)>
  <Container
    className="pt-6 blocks-container"
    tag="div"
  >
    <Row
      tag="div"
    >
      <Col
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
        xs="12"
      >
        <PageTitle
          secondaryTitle=""
          title="explorer.delegates.title"
        />
        <h4>
          <FormattedMessage
            id="explorer.delegates.table-active.title"
            values={Object {}}
          />
        </h4>
        <WitnessesTable
          isCandidates={false}
          witnesses={Array []}
        />
        <h4>
          <FormattedMessage
            id="explorer.delegates.table-candidates.title"
            values={Object {}}
          />
        </h4>
        <WitnessesTable
          isCandidates={true}
          witnesses={Array []}
        />
      </Col>
    </Row>
  </Container>
</Connect(MainUI)>
`;
